File: LeetCode/valid-parenthesis-string/valid-parenthesis-string.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool checkValidString(string s) {
        vector<int> stack, wildcard;

        for (int idx = 0; idx<s.length(); idx++){
            if (s[idx] == '('){
                stack.push_back(idx);
                continue;
            } else if (s[idx] == '*'){
                wildcard.push_back(idx);
                continue;
            }
            if (stack.size() == 0 && wildcard.size() == 0){
                return false;
            }
            if (stack.size() > 0){
                stack.pop_back();
                continue;
            }
            wildcard.pop_back();
        }

        int index = 0;
        for (int i = 0; i<stack.size(); i++){
            bool found = false;
            for (int j = index; j<wildcard.size(); j++){
                if (wildcard[j] > stack[i]){
                    index = j+1;
                    found = true;
                    break;
                }
            }
            if (!found) return false;
        }
        
        return true;
    }
};

int main(){
    Solution solution_;
    
    string s1 = "((((()(()()()*()(((((*)()*(**(())))))(())()())(((())())())))))))(((((())*)))()))(()((*()*(*)))(*)()";
	cout<<solution_.checkValidString(s1)<<endl;
	// true

	string s2 = "(((((*(()((((*((**(((()()*)()()()*((((**)())*)*)))))))(())(()))())((*()()(((()((()*(())*(()**)()(())";
	cout<<solution_.checkValidString(s2)<<endl;
    // false

    return 0;
}

File: LeetCode/valid-parenthesis-string/valid-parenthesis-string.go
package main

import "fmt"

func checkValidString(s string) bool {
	var (
		stack    []int
		wildcard []int
	)

	for idx, char := range s {
		if char == '(' {
			stack = append(stack, idx)
			continue
		} else if char == '*' {
			wildcard = append(wildcard, idx)
			continue
		}
		if len(stack) == 0 && len(wildcard) == 0 {
			return false
		}
		if len(stack) > 0 {
			stack = stack[:len(stack)-1]
			continue
		}
		wildcard = wildcard[:len(wildcard)-1]
	}

	index := 0

	for i := 0; i < len(stack); i++ {
		found := false
		for j := index; j < len(wildcard); j++ {
			if wildcard[j] > stack[i] {
				index = j + 1
				found = true
				break
			}
		}
		if !found {
			return false
		}
	}

	return true
}

func main() {
	var s1 = "((((()(()()()*()(((((*)()*(**(())))))(())()())(((())())())))))))(((((())*)))()))(()((*()*(*)))(*)()"
	fmt.Println(checkValidString(s1))
	// true

	var s2 = "(((((*(()((((*((**(((()()*)()()()*((((**)())*)*)))))))(())(()))())((*()()(((()((()*(())*(()**)()(())"
	fmt.Println(checkValidString(s2))
	// false
}
